lib/Onyx.js

```diff
@@ -1034,26 +1034,31 @@ function initializeWithDefaultKeyStates() {
 function clear(keysToPreserve = []) {
     return getAllKeys()
         .then((keys) => {
+            const keyValuesToClear = [];
             const defaultKeys = _.keys(defaultKeyStates);
 
-            // We're clearing out (setting to null) all keys aren't being preserved or have defaults
-            const keysToClear = _.without(keys, ..._.union(keysToPreserve, defaultKeys));
-            const keyValuesToClear = _.map(keysToClear, key => [key, null]);
+            // The only keys that should be cleared are:
+            // 1. Anything specifically passed in keysToPreserve (because some keys like language preferences, offline status, or activeClients need to remain in Onyx even when signed out)
+            // 2. Any keys with a default state (because they need to remain in Onyx as their default, and setting them to null would cause unknown behavior)
+            const keysToClear = _.difference(keys, keysToPreserve, defaultKeys);
+            keyValuesToClear.push(..._.map(keysToClear, key => [key, null]));
 
-            // If keysToPreserve contains a key that also has a default key state, we'll preserve it rather than reset
-            // it to the default.
-            const defaultKeyValues = _.pairs(_.omit(defaultKeyStates, ...keysToPreserve));
+            // Remove any keysToPreserve from the defaultKeyStates because if they are passed in it has been explicitly called out to preserve those values instead of resetting them back
+            // to the default.
+            const defaultKeyValuePairs = _.pairs(_.omit(defaultKeyStates, ...keysToPreserve));
 
-            // Combine the pairs of keys we're clearing and keys we're setting back to default for our multiSet call
-            const newStoreKeyValues = _.union(keyValuesToClear, defaultKeyValues);
+            // Add the default key value pairs to the keyValuesToClear so that they get set back to their default values when we clear Onyx
+            keyValuesToClear.push(...defaultKeyValuePairs);
```

Comment:
**Nitpick:** what about `keyValuesToReset`? We aren't clearing the defaultKeys we are actually resetting them.

Reply:
I like!